File: Listas y pilas/Pila/c++/pila con arreglo/pilaconlista.cpp
#include <iostream>
#include <stdexcept>

template <typename T>
class Pila
{
private:
    struct Nodo
    {
        T valor;
        Nodo* siguiente;

        Nodo(T v) : valor(v), siguiente(nullptr) {}
    };

    Nodo* cima;
    int contador;

public:
    Pila() : cima(nullptr), contador(0) {}

    ~Pila()
    {
        while (cima != nullptr)
        {
            Nodo* temp = cima;
            cima = cima->siguiente;
            delete temp;
        }
    }

    void apilar(T valor)
    {
        Nodo* nuevoNodo = new Nodo(valor);
        nuevoNodo->siguiente = cima;
        cima = nuevoNodo;
        contador++;
    }

    T desapilar()
    {
        if (estaVacia())
        {
            throw std::runtime_error("La pila está vacía (Stack Underflow).");
        }
        
        T valor = cima->valor;
        Nodo* temp = cima;
        cima = cima->siguiente;
        delete temp;
        contador--;
        return valor;
    }

    T verCima()
    {
        if (estaVacia())
        {
            throw std::runtime_error("La pila está vacía.");
        }
        return cima->valor;
    }

    bool estaVacia()
    {
        return cima == nullptr;
    }

    int tamano()
    {
        return contador;
    }
};

File: Listas y pilas/Pila/c#/pila con lista.cs
using System;

public class Pila<T>
{
    private class Nodo
    {
        public T Valor { get; set; }
        public Nodo Siguiente { get; set; }

        public Nodo(T valor)
        {
            Valor = valor;
            Siguiente = null;
        }
    }

    private Nodo cima;
    private int contador;

    public Pila()
    {
        cima = null;
        contador = 0;
    }

    public void Apilar(T item)
    {
        Nodo nuevoNodo = new Nodo(item);
        nuevoNodo.Siguiente = cima;
        cima = nuevoNodo;
        contador++;
    }

    public T Desapilar()
    {
        if (EstaVacia())
        {
            throw new InvalidOperationException("La pila está vacía (Stack Underflow).");
        }
        
        T item = cima.Valor;
        cima = cima.Siguiente;
        contador--;
        return item;
    }

    public T VerCima()
    {
        if (EstaVacia())
        {
            throw new InvalidOperationException("La pila está vacía.");
        }
        return cima.Valor;
    }

    public bool EstaVacia()
    {
        return contador == 0;
        // o también: return cima == null;
    }

    public int Tamano()
    {
        return contador;
    }
}